### ISSUE TITLE
Using exact Hessian for multinomial logistic

### ISSUE THREAD

**Issue by david-cortes** [dmlc/xgboost#12278]
When using multi-output trees for multinomial logistic objective (`multi:softprob` and `multi:softmax`), the objective is treated the same way as multi-task regression, where the tasks have no dependencies between them, but in multinomial logistic's case, the true Hessians (second derivative) are $k \times k$ matrices which have non-zero values in the off-diagonal elements - i.e. tasks are not independent.

XGBoost uses a diagonal upper bound of the Hessians instead, which has much poorer convergence and results in suboptimal leaf values and thereby poor results.

Quick example using this diagonal upper bound to estimate intercepts from a sample of data - takes over 400 Newton iterations to converge to the correct result, while if using the true Hessian, it would take only 4 iterations:

<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/e388abe9-bdf4-40e9-af4d-99a5a3df4052" />

Link to the code:
https://github.com/david-cortes/vector_gbt_poc/blob/master/Example1_Calculating_Intercepts.ipynb

Here I wrote a quick article describing the mathematics behind gradient boosting with vector functions and what the procedure could look like in the case of multinomial logistic:
https://arxiv.org/pdf/2606.29326

Would be better if XGBoost could use the actual true Hessian for multinomial logistic and for custom objectives instead. Procedure explained in the article would be similar as the current one (quick Python POC: https://github.com/david-cortes/vector_gbt_poc/blob/master/Example3_Split_Point_and_Leaf_Calculation.ipynb), just with involvement of some Cholesky factorizations and linear solves. Plus it'd make it easier to add more objectives involving vector functions in the future.

Can provide help with code reviews and answering questions about the theoretical parts or about the methodology.

Note: the procedure requires the sums of Hessians to be accurate enough that they end up being positive-definite (which they are by definition, if using real numbers) or close enough if regularized. Not sure how float32 or quantized numbers would fare in this regard though.

CC @hcho3 @RAMitchell

**Comment by trivialfis** [comment]
> just with involvement of some Cholesky factorizations and linear solves. 

That's part of the reason we don't support it. ;-)

That said, if convergence is faster, we can look into it.

**Comment by RAMitchell** [comment]
Two relevant papers:
https://arxiv.org/abs/1203.3491
https://arxiv.org/abs/1110.3907
https://arxiv.org/abs/1110.3907

This is relatively well known. As @trivialfis said its not really practical. The linear solves are both extremely expensive ( $\mathcal{O}(n^3)$ ) and numerically unstable - it can just fail. We also blow out memory for large numbers of classes.

In one tree-per-class mode, one idea would be to recompute the gradients after each tree, instead of after each group of trees. This turns it into a coordinate descent algorithm which is much better at dealing with curvature.

**Comment by david-cortes** [comment]
> That said, if convergence is faster, we can look into it.

Picture in the first message shows it takes several dozen iterations of diagonal-upper-bounded Newton steps to reach the same log-loss as a single exact-Newton iteration on the simplest kind of problem.

> That's part of the reason we don't support it. ;-)

What would be the problem with adding a procedure with Cholesky factorizations and linear solves?

> The linear solves are both extremely expensive ( O ( n 3 ) )

It is indeed an O(n^3) procedure, but 'n' is typically between 3 and 15. For something typical like n=5 and 255 possible splint points, that procedure is likely going to complete in less than a millisecond.

> and numerically unstable - it can just fail.

It _might_ be numerically unstable in some corner cases (remark that it'd be used for one particular form of Hessian which is 'diagonal plus rank-k'), but lots of libraries use Cholesky factorization successfully for Newton iterations of multinomial logistic likelihoods (for example, scikit-learn). Even then, if it fails, there's still other ways besides Cholesky, like coordinate descent (O(n^2)), or pseudo inverses (O(n^3)).

> We also blow out memory for large numbers of classes.

The amount of extra memory would amount to something in the order of `3*n^2*nthreads` (not sure if xgboost keeps histograms at each node - that would increase a bit more). Assuming a big problem like n=15 and 64 threads, that's less than a megabyte.

> This turns it into a coordinate descent algorithm which is much better at dealing with curvature.

That kind of algorithm also has slower convergence. It might not be as bad as diagonal upper bounds, but you might find that it similarly takes much more iterations of a procedure like that to match a single exact Newton iteration.

**Comment by trivialfis** [comment]
We need to store the entire gradient matrix for every sample. Turning a sample into a matrix can be problematic real quick: `n=15, n^2=225`, with each element taking 8 bytes (grad, hess). Even if it's symmetric, it's still a lot of memory. Yes, the histogram is stored for each node for efficient split calculation.

As for the matrix inverse, we will either add dependencies or implement it ourselves. This is probably useful if we were to expand to other types of trees with embedded linear models, that is another story.

Overall, it's a tradeoff we made. I'm not against of making the trade off optional if there's a real use case, but it needs to be compelling. To actually make it usable for large dataset, we might have to treat the gradient similar to how we treat the feature matrix when it comes to storage, which is not trivial. You worked with the DMatrix iterator before.

**Comment by david-cortes** [comment]
> We need to store the entire gradient matrix for every sample. Turning a sample into a matrix can be problematic real quick: `n=15, n^2=225`, with each element taking 8 bytes (grad, hess). Even if it's symmetric, it's still a lot of memory. Yes, the histogram is stored for each node for efficient split calculation.

In the case of multinomial logistic, if viewed in terms of predicted probabilities, the hessian can be calculated as:
$$\textbf{diag}(\mathbf{p}) + \mathbf{p} \mathbf{p}^T$$

And the gradient as:
$$\mathbf{y} - \mathbf{p}$$

Thus, you only need to calculate the row-wise predicted probabilities, and from that can make up the aggregations on the fly. Only the aggregations need O(n^2) elements.

> As for the matrix inverse, we will either add dependencies or implement it ourselves. This is probably useful if we were to expand to other types of trees with embedded linear models, that is another story.

Not sure about all the language bindings, but for Python at least, XGBoost already depends on SciPy which provides bindings for LAPACK for C/C++ code which includes all of those:
https://github.com/dmlc/xgboost/blob/811dadfadd4fd11adc9f5dec9c9afba9e649361c/python-package/pyproject.toml#L34

and on R, LAPACK is a built-in. Either way, it's not hard to create an inefficient textbook Cholesky or CD solver if it's only going to be applied to small problems like n=15. Plus there's header-only libraries like Eigen (which lightgbm uses for this purpose for example, on linear tree leafs).

> Overall, it's a tradeoff we made. I'm not against of making the trade off optional if there's a real use case, but it needs to be compelling. To actually make it usable for large dataset, we might have to treat the gradient similar to how we treat the feature matrix when it comes to storage, which is not trivial. You worked with the DMatrix iterator before.

Yes, it would require a different procedure indeed, producing a kind of "compressed" hessian.
